From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: Reilly Brogan <reilly@reillybrogan.com>
Date: Sun, 26 May 2024 16:08:54 -0500
Subject: [PATCH] Emit relocations for BOLT

Ensure that relocations are always emitted for the libraries/binaries that we want to BOLT. This allows us to not have them emitted for all other files.

Origin: vendor
Forwarded: not-needed
Last-Update: 2025-03-29
---
 clang/tools/clang-shlib/CMakeLists.txt | 2 ++
 clang/tools/driver/CMakeLists.txt      | 2 ++
 clang/tools/libclang/CMakeLists.txt    | 3 +++
 lld/Common/CMakeLists.txt              | 2 ++
 lld/ELF/CMakeLists.txt                 | 2 ++
 lld/tools/lld/CMakeLists.txt           | 2 ++
 llvm/tools/llvm-shlib/CMakeLists.txt   | 2 ++
 llvm/tools/lto/CMakeLists.txt          | 2 ++
 polly/lib/CMakeLists.txt               | 2 ++
 polly/lib/External/CMakeLists.txt      | 2 ++
 10 files changed, 21 insertions(+)

diff --git a/clang/tools/clang-shlib/CMakeLists.txt b/clang/tools/clang-shlib/CMakeLists.txt
index 945076e1ad81..8e25db254ac4 100644
--- a/clang/tools/clang-shlib/CMakeLists.txt
+++ b/clang/tools/clang-shlib/CMakeLists.txt
@@ -60,6 +60,8 @@ endif()
 if (NOT APPLE AND LLVM_LINKER_SUPPORTS_B_SYMBOLIC_FUNCTIONS)
   target_link_options(clang-cpp PRIVATE LINKER:-Bsymbolic-functions)
 endif()
+# We need to emit relocations to BOLT this library on Solus
+target_link_options(clang-cpp PRIVATE LINKER:--emit-relocs)
 if (MINGW OR CYGWIN)
   # The clang-cpp DLL is supposed to export all symbols (except for ones
   # that are explicitly hidden). Normally, this is what happens anyway, but
diff --git a/clang/tools/driver/CMakeLists.txt b/clang/tools/driver/CMakeLists.txt
index d9d36f7a4135..17181f024935 100644
--- a/clang/tools/driver/CMakeLists.txt
+++ b/clang/tools/driver/CMakeLists.txt
@@ -211,3 +211,5 @@ if (CLANG_BOLT AND NOT LLVM_BUILD_INSTRUMENTED)
     VERBATIM
   )
 endif()
+
+target_link_options(clang PRIVATE LINKER:--emit-relocs)
diff --git a/clang/tools/libclang/CMakeLists.txt b/clang/tools/libclang/CMakeLists.txt
index 2b1e266f0739..f6f0e87ba3c4 100644
--- a/clang/tools/libclang/CMakeLists.txt
+++ b/clang/tools/libclang/CMakeLists.txt
@@ -210,6 +210,9 @@ if(ENABLE_SHARED)
   endif()
 endif()
 
+# We need to emit relocations to BOLT this library
+target_link_options(libclang PRIVATE LINKER:--emit-relocs)
+
 if(INTERNAL_INSTALL_PREFIX)
   set(LIBCLANG_HEADERS_INSTALL_DESTINATION "${INTERNAL_INSTALL_PREFIX}/include")
 else()
diff --git a/lld/Common/CMakeLists.txt b/lld/Common/CMakeLists.txt
index a9e8d72fb5ec..b640b44ec406 100644
--- a/lld/Common/CMakeLists.txt
+++ b/lld/Common/CMakeLists.txt
@@ -59,3 +59,5 @@ add_lld_library(lldCommon
   DEPENDS
   intrinsics_gen
   )
+
+target_link_options(lldCommon PRIVATE LINKER:--emit-relocs)
diff --git a/lld/ELF/CMakeLists.txt b/lld/ELF/CMakeLists.txt
index ec3f6382282b..cb0a7dbc46db 100644
--- a/lld/ELF/CMakeLists.txt
+++ b/lld/ELF/CMakeLists.txt
@@ -87,3 +87,5 @@ add_lld_library(lldELF
   ELFOptionsTableGen
   intrinsics_gen
   )
+
+target_link_options(lldELF PRIVATE LINKER:--emit-relocs)
diff --git a/lld/tools/lld/CMakeLists.txt b/lld/tools/lld/CMakeLists.txt
index 8498a91597a9..3e7b95c3953a 100644
--- a/lld/tools/lld/CMakeLists.txt
+++ b/lld/tools/lld/CMakeLists.txt
@@ -46,3 +46,5 @@ endforeach()
 if(LLVM_TOOL_LLVM_DRIVER_BUILD)
   set_property(GLOBAL APPEND PROPERTY LLVM_DRIVER_HIDDEN_TOOL_ALIASES_lld ld)
 endif()
+
+target_link_options(lld PRIVATE LINKER:--emit-relocs)
diff --git a/llvm/tools/llvm-shlib/CMakeLists.txt b/llvm/tools/llvm-shlib/CMakeLists.txt
index 53003d90160f..b233aa259cfc 100644
--- a/llvm/tools/llvm-shlib/CMakeLists.txt
+++ b/llvm/tools/llvm-shlib/CMakeLists.txt
@@ -85,6 +85,8 @@ if(LLVM_BUILD_LLVM_DYLIB)
         # Note: for -fno-pic default, the address of a function may be different from
         # inside and outside libLLVM.so.
         target_link_options(LLVM PRIVATE LINKER:-Bsymbolic-functions)
+        # We need to emit relocations to BOLT this library on Solus/Aeryn
+        target_link_options(LLVM PRIVATE LINKER:--emit-relocs)
       endif()
     endif()
   endif()
diff --git a/llvm/tools/lto/CMakeLists.txt b/llvm/tools/lto/CMakeLists.txt
index 8c7d7904e16e..4251ec8423b6 100644
--- a/llvm/tools/lto/CMakeLists.txt
+++ b/llvm/tools/lto/CMakeLists.txt
@@ -36,6 +36,8 @@ if(LLVM_ENABLE_PIC)
   add_llvm_library(${LTO_LIBRARY_NAME} ${LTO_LIBRARY_TYPE} INSTALL_WITH_TOOLCHAIN
       ${SOURCES} DEPENDS intrinsics_gen)
 
+  target_link_options(${LTO_LIBRARY_NAME} PRIVATE LINKER:--emit-relocs)
+
   install(FILES ${LLVM_MAIN_INCLUDE_DIR}/llvm-c/lto.h
     DESTINATION "${CMAKE_INSTALL_INCLUDEDIR}/llvm-c"
     COMPONENT LTO)
diff --git a/polly/lib/CMakeLists.txt b/polly/lib/CMakeLists.txt
index 965f635b7ff6..77c0be2fc4b2 100644
--- a/polly/lib/CMakeLists.txt
+++ b/polly/lib/CMakeLists.txt
@@ -132,6 +132,8 @@ else ()
   # the same command line options, to which LLVM reacts with an error.
   target_link_libraries(LLVMPolly PUBLIC ${ISL_TARGET})
 
+  target_link_options(LLVMPolly PRIVATE LINKER:--emit-relocs)
+
   set_target_properties(LLVMPolly
     PROPERTIES
     LINKER_LANGUAGE CXX
diff --git a/polly/lib/External/CMakeLists.txt b/polly/lib/External/CMakeLists.txt
index f065fbd7b942..299a4c06febb 100644
--- a/polly/lib/External/CMakeLists.txt
+++ b/polly/lib/External/CMakeLists.txt
@@ -288,6 +288,8 @@ if (POLLY_BUNDLED_ISL)
     ${ISL_FILES}
     )
 
+  target_link_options(PollyISL PRIVATE LINKER:--emit-relocs)
+
 
   if (NOT LLVM_INSTALL_TOOLCHAIN_ONLY)
     install(DIRECTORY
